File: quick-sort/quick-sort.rb
module QuickSort
  extend self
  require 'wikipedia'

  def sort(array: [], low: 0, high: array.length - 1, order: :asc)
    if low < high
      p_index = partition(array, low, high, order)
      sort(array: array, low: low, high: p_index - 1, order: order)
      sort(array: array, low: p_index + 1, high: high, order: order)
    end

    array
  end

  def partition(array, low, high, order)
    pivot   = array[high]
    p_index = low

    if order == :asc
      for i in low..(high - 1)
        if array[i] < pivot
          swap(array, i, p_index)
          p_index += 1
        end
      end
    end

    if order == :desc
      for i in low..(high - 1)
        if array[i] > pivot
          swap(array, i, p_index)
          p_index += 1
        end
      end
    end

    swap(array, p_index, high)

    p_index
  end

  def swap(array, a, b)
    temp = array[a]
    array[a] = array[b]
    array[b] = temp
  end

  def profile
    page = Wikipedia.find( 'Quick Sort' )
    puts page.summary.split("\n").map { |a| a.strip }.join(" ").gsub(/{.+}/, "")
  end
end

File: quick-sort/java/QuickSort.java
import java.util.Arrays;

public class QuickSort {

  public enum Order { ASC, DESC }

  public static int[] sort( int[] array ) {
    return sort(array, Order.ASC);
  }

  public static int[] sort( int[] array, Order order ) {
    int low = 0;
    int high = array.length - 1;
    int[] result;

    return sort(array, low, high, order);
  }

  public static int[] sort( int[] array, int low, int high, Order order ) {
    if (low < high) {
      int pIndex = partition(array, low, high, order);
      sort(array, low, pIndex - 1, order);
      sort(array, pIndex + 1, high, order);
    }

    return array;
  }

  public static int partition( int[] array, int low, int high, Order order ) {
    int pivot = array[high];
    int pIndex = low;

    if (order == Order.ASC) {
      for (int i = low; i <= high - 1; i++) {
        if (array[i] < pivot) {
          swap(array, i, pIndex);
          pIndex ++;
        }
      }
    }

    if (order == Order.DESC) {
      for (int i = low; i <= high - 1; i++) {
        if (array[i] > pivot) {
          swap(array, i, pIndex);
          pIndex ++;
        }
      }
    }

    swap(array, pIndex, high);

    return pIndex;
  }

  public static int[] swap( int[] array, int index1, int index2 ) {
    int temp = array[index1];
    array[index1] = array[index2];
    array[index2] = temp;
    return array;
  }
}
